modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/construction/ConstructionService.scala

```diff
@@ -106,5 +118,92 @@ object ConstructionService {
         parseUnsignedTransaction(hex)
       }
+
+    def combineTransaction(hex: Hex, signatures: NonEmptyList[RosettaSignature]): EitherT[F, Throwable, Either[ConstructionError, Hex]] =
+      KryoSerializer[F]
+        .deserialize[Transaction](hex.toBytes)
+        .traverse { transaction =>
+          convertSignature(signatures).map { test2 =>
+            test2.left
+              .map(err => Left(err): Either[ConstructionError, Hex])
+              .map { signatureProofs =>
+                val serializedTransaction: Either[ConstructionError, Hex] = KryoSerializer[F]
+                  .serialize(
+                    Signed[Transaction](
+                      transaction,
+                      NonEmptySet(signatureProofs.head, SortedSet(signatureProofs.tail: _*))(
+                        Order.fromOrdering(SignatureProof.OrderingInstance)
+                      )
+                    )
+                  )
+                  .left
+                  .map(_ => MalformedTransaction)
+                  .map(bytes => Hex.fromBytes(bytes))
+
+                serializedTransaction
+              }
+              .merge
+          }
+        }
+        .map { test =>
+          test.left.map(_ => Left(MalformedTransaction): Either[ConstructionError, Hex]).merge
+        }
+        .attemptT
+
+    def convertSignature(
+      signature: NonEmptyList[RosettaSignature]
+    ) = {
+      val value = signature.map { s =>
+        val value1 = Async[F].delay(getPublicKeyFromBytes(s.publicKey.hexBytes.toBytes))
+        value1.map { pk =>
+          s.signatureType match {
+            case SignatureType.Ecdsa =>
+              Right {
+                SignatureProof(pk.toId, org.tessellation.security.signature.signature.Signature(s.hexBytes))
+              }
+            case x => Left(UnsupportedOperation)
+          }
+        }
+      }
+      val zero: Either[ConstructionError, List[SignatureProof]] = Right[ConstructionError, List[SignatureProof]](List())
+      value
+        .foldLeft(Async[F].delay(zero)) {
+          case (agga, nextFEither) =>
+            val inner2 = agga.flatMap { agg =>
+              val inner = nextFEither.map { eitherErrSigProof =>
+                val res = eitherErrSigProof.map { sp =>
+                  agg.map(ls => ls ++ List(sp))
+                }
+                var rett: Either[ConstructionError, List[SignatureProof]] = null
+                if (res.isRight) {
+                  val value2 = res.toOption.get
+                  val zz = value2.getOrElse(List())
+                  if (zz.nonEmpty) {
+                    rett = Right(zz)
+                  } else {
+                    rett = Left(value2.swap.toOption.get)
+                  }
+                } else {
+                  rett = Left(res.swap.toOption.get)
+                }
+                rett
+              }
+              inner
+            }
+            inner2
+        }
+        .map(_.map(q => NonEmptyList(q.head, q.tail)))
```

Comment:
it's very error prone and actually obfuscated what it does. if you start doing things such way, think twice or ask, because most likely it can be written in way simpler

Reply:
Can you elucidate on what is bad here?

Reply:
First of all, you use mutations which should be avoided as much as possible.
Secondly, there are a lot of auxiliary variables that are not necessary. Converting from Option to Either is overengineered, you don't need to use `if`. You used `.get` operation on option which is unsafe and can throw in the runtime. Overall, it looks very imperative and it becomes unmaintainable really quickly. 

There are several reasons why declarative programming is often considered better and easier to understand than imperative programming:

- Readability: Declarative code is often more readable because it focuses on the intent of the code rather than the implementation details. This makes it easier for other developers to understand and maintain the code.

- Simplicity: Declarative code is often simpler and more concise than imperative code because it abstracts away many of the low-level details. This can lead to faster development times and fewer bugs.

- Reusability: Declarative code is often more reusable than imperative code because it can be used in a wider variety of contexts. This can lead to more modular and maintainable code.

- Parallelism: Declarative programming is often more amenable to parallelization because it specifies the desired outcome rather than the exact steps needed to achieve it. This can lead to faster execution times on multi-core or distributed systems.

Overall, declarative programming is often seen as a more natural and intuitive way of expressing complex computations because it allows the programmer to focus on the what rather than the how.

Instead of thinking **how** to do something, start thinking **what** to do.

**As an example:**

I have a non empty list of signatures. I want to convert each signature from rosetta schema to SignatureProof. 

```scala
    def convertSignatures(signatures: NonEmptyList[RosettaSignature]): EitherT[F, ConstructionError, NonEmptyList[SignatureProof]] = ???
```

I know that for each signature, I need to derive a public key:

```scala
signatures.traverse { signature =>
         signature.publicKey.hexBytes.toPublicKeyByEC.map { pk =>
         
         }
}
```

then, based on the rosetta specification, I want only ECDSA signatures to be converted:

```scala
            signature.signatureType match {
              case SignatureType.Ecdsa => ???
              case _                   => ???
            }
```

For ECDSA signature type, I want to create a SignatureProof, for other cases I want to return an unsupported operation.

```scala
         case SignatureType.Ecdsa => SignatureProof(pk.toId, Signature(signature.hexBytes)).asRight[ConstructionError]
         case _                   => UnsupportedOperation.asInstanceOf[ConstructionError].asLeft[SignatureProof]
```

I return it wrapped into an Either (to have a common type), so I end up with type:

```scala
F[NonEmptyList[Either[ConstructionError, Hex]]]
```

I need sequence operation to convert from `F[G[A]]` into `G[F[A]]`
```scala
.map(_.sequence)
```

so it becomes:
```scala
F[Either[ConstructionError, NonEmptyList[Hex]]]
```

Then, I want to convert `F[Either` into `EitherT[F`, so I need to wrap it into `EitherT { ... }`.
Done.

Reply:
No, traverse **is** the map then sequence. (Take a look info docs and traverse/sequence implementations)

Reply:
or actually @cngo-github is the opposite, `sequence` **is** the `traverse` but with identity map function:
![image](https://user-images.githubusercontent.com/4959931/225120809-cd824a7f-1d76-49e0-8d47-8d2369c3abe9.png)
![image](https://user-images.githubusercontent.com/4959931/225120836-842bf45a-67b6-452c-a4b5-ad04183b2d53.png)

Reply:
also this @cngo-github 
![image](https://user-images.githubusercontent.com/4959931/225122404-6a20014e-5ed9-4a94-8e0d-82425eb33744.png)